mlir/lib/Dialect/Tosa/Transforms/TosaInferShapes.cpp

```diff
@@ -39,9 +34,87 @@ using namespace mlir::tosa;
 
 namespace {
 
-void propagateShapesInRegion(Region &region);
+// Check whether this use case is replaceable. We define an op as
+// being replaceable if it is used by a TosaOp, or an op with a
+// type-inference related interface.
+// When a non-replaceable use is encountered, the value is wrapped in a
+// cast back to the original type after inference.
+bool isReplaceableUser(Operation *user) {
+  // Handle unregistered dialects.
+  if (!user->getDialect())
+    return false;
+
+  return user->getDialect()->getNamespace() ==
```

Comment:
Note: this is a string compare for dialects, comparing the id's would be cheaper.

Reply:
This code already existed in the pass. I just moved it into a separate function so it could be used elsewhere.